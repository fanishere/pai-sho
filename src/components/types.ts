export type Position = {
  x: number;
  y: number;
};

export type PieceData = {
  position: Position;
};

export interface PieceDataList {
  [index: string]: PieceData;
}

export type PieceGuideCandidates = {
  position: Position;
  offset: number;
};

export type PieceGuide = {
  position: Position;
  offset: number;
};

export type Side = 'red' | 'white';
