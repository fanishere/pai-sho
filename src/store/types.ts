export interface Position {
  x: number;
  y: number;
}

export interface PieceData {
  name: string;
  position: Position | null;
}

export type PieceGuide = {
  position: Position;
  offset: number;
};

export interface GameState {
  piecePositions: PieceData[];
  guidesVisible: PieceGuide[];
}

export const MOVE_PIECE = 'MOVE_PIECE';
export const SET_GUIDES = 'SET_GUIDES';
export const REMOVE_GUIDES = 'REMOVE_GUIDES';

export interface PieceMoveAction {
  type: typeof MOVE_PIECE;
  pieceName: string;
  position: Position;
}

export interface SetGuidesAction {
  type: typeof SET_GUIDES;
  guides: PieceGuide[];
}

export interface RemoveGuidesAction {
  type: typeof REMOVE_GUIDES;
}

export type GameActionTypes =
  | PieceMoveAction
  | SetGuidesAction
  | RemoveGuidesAction;
