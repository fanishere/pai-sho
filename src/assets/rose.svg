<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 20010904//EN"
 "http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd">
<svg version="1.0" xmlns="http://www.w3.org/2000/svg"
 width="602.000000pt" height="602.000000pt" viewBox="0 0 602.000000 602.000000"
 preserveAspectRatio="xMidYMid meet">
<metadata>
Created by potrace 1.16, written by Peter Selinger 2001-2019
</metadata>
<g transform="translate(0.000000,602.000000) scale(0.100000,-0.100000)"
fill="#000000" stroke="none">
<path d="M2820 6014 c-19 -2 -84 -9 -145 -15 -236 -23 -531 -95 -766 -188
-392 -154 -719 -369 -1025 -675 -483 -484 -769 -1066 -861 -1753 -23 -177 -23
-569 0 -746 92 -687 378 -1269 861 -1753 484 -483 1066 -769 1753 -861 177
-23 569 -23 746 0 687 92 1269 378 1753 861 483 484 769 1066 861 1753 23 177
23 569 0 746 -92 687 -378 1269 -861 1753 -482 482 -1074 773 -1747 859 -117
15 -494 28 -569 19z m615 -44 c668 -101 1250 -398 1716 -876 446 -459 719
-1012 820 -1664 20 -124 23 -184 23 -420 0 -302 -15 -438 -74 -685 -264 -1106
-1119 -1961 -2225 -2225 -247 -59 -383 -74 -685 -74 -236 0 -296 3 -420 23
-652 101 -1205 374 -1664 820 -479 467 -772 1042 -877 1721 -20 124 -23 184
-23 420 0 236 3 296 23 420 105 679 400 1257 878 1721 464 450 1033 730 1663
818 203 29 208 29 470 26 176 -3 270 -9 375 -25z"/>
<path d="M2948 5704 c-62 -33 -80 -147 -33 -203 34 -40 84 -56 130 -41 54 18
85 62 85 121 0 28 -7 63 -15 79 -28 55 -109 76 -167 44z m130 -33 c36 -37 46
-81 28 -124 -20 -48 -52 -72 -96 -72 -44 0 -76 24 -96 72 -30 69 23 153 96
153 29 0 46 -7 68 -29z"/>
<path d="M2471 5512 c-79 -12 -126 -115 -86 -192 62 -120 235 -75 235 60 0 86
-66 144 -149 132z m63 -27 c104 -49 73 -215 -40 -215 -51 0 -104 53 -104 104
0 50 23 90 62 109 41 20 43 20 82 2z"/>
<path d="M3483 5509 c-51 -15 -83 -61 -83 -120 0 -89 46 -139 127 -139 71 0
113 52 113 139 0 71 -34 110 -110 126 -8 2 -29 -1 -47 -6z m85 -29 c46 -28 68
-82 53 -131 -32 -107 -172 -105 -201 1 -8 31 -7 47 4 74 15 35 67 76 96 76 8
0 30 -9 48 -20z"/>
<path d="M3000 5376 c0 -2 -9 -102 -20 -222 -68 -757 -78 -1433 -25 -1733 15
-89 45 -191 55 -191 10 0 40 102 55 191 53 300 43 976 -25 1733 -11 120 -20
220 -20 222 0 2 -4 4 -10 4 -5 0 -10 -2 -10 -4z m34 -363 c9 -120 22 -288 28
-373 17 -222 17 -999 0 -1120 -18 -124 -41 -230 -52 -230 -11 0 -34 106 -52
230 -17 121 -17 898 0 1120 36 477 46 590 52 590 4 0 14 -98 24 -217z"/>
<path d="M2013 5209 c-17 -5 -43 -23 -57 -39 -72 -86 -19 -221 87 -220 78 1
127 51 127 130 0 94 -71 153 -157 129z m77 -24 c99 -51 68 -214 -41 -215 -97
0 -143 125 -71 192 44 41 68 46 112 23z"/>
<path d="M3936 5209 c-137 -39 -114 -260 27 -259 78 1 127 51 127 130 0 93
-72 153 -154 129z m74 -24 c99 -51 68 -214 -41 -215 -97 0 -143 125 -71 192
44 41 68 46 112 23z"/>
<path d="M2544 4943 c18 -646 68 -1142 146 -1451 28 -108 85 -252 100 -252 18
0 35 143 34 290 -1 170 -13 298 -50 555 -57 389 -204 1095 -227 1095 -7 0 -8
-87 -3 -237z m71 -107 c95 -437 162 -847 186 -1136 11 -142 7 -337 -9 -394 -7
-27 -9 -26 -35 40 -93 241 -154 679 -187 1353 -10 203 -12 375 -4 356 2 -5 24
-104 49 -219z"/>
<path d="M3446 5108 c-100 -415 -201 -971 -237 -1308 -24 -224 -11 -560 21
-560 15 0 72 144 100 252 78 309 128 805 146 1451 5 150 4 237 -3 237 -5 0
-17 -33 -27 -72z m4 -388 c-31 -682 -92 -1129 -187 -1374 -26 -66 -28 -67 -35
-40 -16 57 -20 252 -9 394 24 293 85 666 183 1120 63 289 66 282 48 -100z"/>
<path d="M3232 5022 c-60 -259 -97 -541 -89 -686 7 -148 20 -169 55 -88 46
111 62 240 69 580 6 308 -1 346 -35 194z m15 -259 c-5 -227 -21 -368 -52 -472
-12 -39 -19 -50 -26 -40 -14 22 -10 216 6 346 18 139 65 393 72 393 3 0 3
-102 0 -227z"/>
<path d="M2753 4823 c9 -343 32 -522 80 -618 l20 -40 13 30 c33 76 10 390 -51
692 -34 169 -44 203 -57 203 -8 0 -9 -73 -5 -267z m88 -198 c29 -199 32 -431
4 -403 -11 10 -43 157 -54 243 -7 50 -14 189 -16 310 l-5 220 25 -110 c13 -60
34 -177 46 -260z"/>
<path d="M3642 4810 c-69 -185 -127 -378 -152 -499 -21 -106 -27 -251 -10
-251 20 0 69 92 99 188 38 118 82 328 110 518 35 235 26 242 -47 44z m23 -81
c-39 -244 -92 -457 -142 -572 -44 -99 -51 -43 -16 138 25 130 158 553 170 541
2 -2 -3 -50 -12 -107z"/>
<path d="M2315 4874 c17 -126 57 -362 75 -446 42 -188 103 -353 143 -387 13
-10 16 -6 22 27 18 107 -65 438 -193 772 -35 90 -57 105 -47 34z m120 -307
c58 -177 105 -384 105 -458 l-1 -44 -24 40 c-56 90 -127 381 -170 695 -5 31
-3 30 15 -15 11 -27 45 -126 75 -218z"/>
<path d="M2100 4876 c0 -31 43 -333 71 -501 79 -470 168 -805 264 -994 32 -64
75 -115 87 -103 15 15 11 219 -7 327 -33 206 -116 507 -233 840 -122 350 -182
492 -182 431z m145 -388 c184 -514 298 -997 270 -1145 -7 -36 -7 -37 -22 -18
-63 82 -153 333 -213 595 -56 238 -152 796 -149 860 1 19 33 -64 114 -292z"/>
<path d="M3850 4753 c-142 -370 -274 -791 -326 -1038 -27 -129 -45 -391 -30
-431 15 -40 84 63 146 220 69 174 146 498 209 871 28 168 71 470 71 501 0 40
-23 0 -70 -123z m15 -159 c-57 -388 -132 -744 -204 -964 -39 -119 -102 -262
-134 -305 -15 -19 -16 -18 -23 19 -16 87 32 369 107 636 72 256 269 825 277
802 1 -5 -9 -90 -23 -188z"/>
<path d="M1350 3560 c-321 -83 -565 -370 -636 -749 -25 -131 -15 -366 20 -484
102 -341 362 -546 716 -564 129 -6 203 7 300 54 67 33 158 122 188 185 26 52
53 171 41 176 -4 2 -28 -13 -54 -32 -122 -93 -285 -158 -400 -159 -186 -1
-365 122 -460 316 -96 197 -99 411 -9 605 170 370 594 446 854 154 56 -63 122
-190 140 -270 18 -78 44 -104 111 -109 62 -5 110 15 143 59 39 53 22 185 -44
340 -101 237 -325 428 -567 482 -98 22 -248 20 -343 -4z m405 -34 c270 -92
484 -340 543 -627 15 -71 15 -79 -2 -122 -12 -33 -26 -49 -50 -61 -39 -18 -98
-21 -127 -6 -21 12 -49 60 -49 85 0 32 -53 151 -97 217 -77 117 -203 209 -323
237 -63 15 -186 13 -246 -3 -319 -85 -505 -470 -403 -836 52 -188 185 -345
349 -410 85 -34 218 -39 311 -10 76 23 179 71 239 112 64 44 63 44 55 7 -40
-196 -229 -330 -465 -330 -215 1 -392 73 -541 221 -75 75 -94 100 -137 191
-83 170 -111 333 -92 523 41 408 292 735 634 827 109 29 293 22 401 -15z"/>
<path d="M4314 3565 c-143 -31 -262 -100 -384 -224 -136 -137 -216 -288 -242
-455 -10 -65 -9 -80 6 -119 24 -63 55 -81 139 -82 84 0 96 12 128 118 52 177
166 320 309 391 194 95 408 50 568 -118 242 -255 239 -682 -6 -933 -197 -202
-447 -205 -730 -10 -40 27 -75 47 -78 45 -10 -10 27 -141 53 -190 118 -217
437 -290 745 -170 252 98 417 310 474 608 24 127 15 334 -19 466 -91 349 -340
609 -645 673 -83 18 -237 18 -318 0z m326 -21 c138 -32 320 -145 412 -257 96
-116 169 -258 210 -411 20 -70 22 -106 22 -266 0 -173 -2 -191 -27 -273 -15
-49 -39 -112 -53 -140 -132 -262 -388 -418 -689 -418 -228 0 -408 125 -460
320 l-12 47 73 -48 c166 -111 345 -153 490 -114 277 74 464 390 424 716 -39
327 -268 560 -550 560 -152 0 -255 -45 -370 -160 -85 -85 -126 -155 -165 -282
-29 -98 -48 -118 -108 -118 -132 0 -168 98 -104 285 98 286 304 485 582 560
68 19 245 18 325 -1z"/>
<path d="M2968 3112 c-51 -56 -122 -176 -161 -272 -61 -148 -97 -319 -115
-545 l-8 -90 -32 55 c-62 106 -97 275 -108 522 -8 172 -6 171 -91 34 -123
-198 -229 -459 -272 -671 -31 -149 -46 -375 -31 -481 31 -239 146 -367 349
-390 77 -8 152 -55 234 -144 152 -164 240 -390 254 -652 3 -65 9 -118 12 -118
4 0 12 62 19 137 24 272 96 469 232 632 78 94 138 132 231 147 103 16 177 51
238 112 148 148 190 411 121 754 -43 213 -149 474 -273 674 -85 137 -83 138
-91 -34 -11 -247 -46 -416 -108 -522 l-32 -55 -8 90 c-24 302 -82 515 -190
697 -40 68 -113 158 -128 158 -4 0 -23 -17 -42 -38z m95 -47 c142 -181 247
-525 247 -813 l0 -63 -42 -19 c-24 -11 -67 -25 -95 -31 -29 -7 -53 -16 -53
-22 0 -22 151 25 212 66 42 29 92 125 118 231 18 72 24 122 48 381 l7 80 47
-75 c68 -109 172 -329 211 -450 66 -198 106 -450 93 -580 -29 -279 -174 -457
-391 -477 -61 -5 -75 -11 -129 -52 -156 -119 -281 -351 -320 -595 l-12 -81
-13 70 c-43 228 -144 424 -287 554 -79 72 -128 98 -202 105 -257 25 -375 256
-333 651 25 230 82 427 187 645 60 125 152 285 159 278 2 -2 8 -72 15 -157 24
-329 80 -490 189 -546 63 -31 162 -58 174 -46 5 5 -13 14 -40 20 -26 7 -69 22
-95 33 l-48 20 0 67 c0 210 70 499 162 673 44 82 122 188 138 188 5 0 29 -25
53 -55z"/>
</g>
</svg>
