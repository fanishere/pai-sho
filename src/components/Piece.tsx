import React from 'react';
import lily from '../assets/Lily_final.svg';
import { Image } from 'react-konva';
import { debounce } from 'underscore';
import Konva from 'konva';
import { PieceGuide, PieceGuideCandidates, Position, PieceData } from './types';

type PieceProps = {
  position: Position;
  name: string;
  gridWidth: number;
  onDragMove: (p: Piece) => void;
  onDragEnd: (p: Piece) => void;
};

export class Piece extends React.Component<PieceProps, PieceData> {
  readonly imageRef: React.RefObject<Konva.Image>;
  constructor(props: PieceProps) {
    super(props);

    this.imageRef = React.createRef();
    this.handleChange = debounce(this.handleChange.bind(this), 25, true);
    this.handleMove = this.handleMove.bind(this);
    this.state = {
      position: this.props.position
    };
  }

  // TODO name the positions ex: top left 0
  getSnapLines = (): Position[] => {
    const lines: Position[] = [];
    const vertical = [
      this.props.position.x - this.props.gridWidth,
      this.props.position.x,
      this.props.position.x + this.props.gridWidth
    ];
    const horizontal = [
      this.props.position.y - this.props.gridWidth,
      this.props.position.y,
      this.props.position.y + this.props.gridWidth
    ];
    vertical.map((v) => {
      horizontal.map((h) => {
        lines.push({
          x: v,
          y: h
        });
      });
    });

    return lines;
  };

  getGuides = (): PieceGuide[] => {
    const GUIDELINE_OFFSET = 100;
    const candidates: PieceGuideCandidates[] = [];
    const offset = 15;

    const snapPositions = this.getSnapLines();
    snapPositions.forEach((pos) => {
      if (this.imageRef.current) {
        const currentPosition = this.imageRef.current.getPosition();
        const xDiff = Math.abs(pos.x - currentPosition.x);
        const yDiff = Math.abs(pos.y - currentPosition.y);

        // if the distance between guild line and
        // object snap point is close we can consider this for snapping
        if (xDiff < GUIDELINE_OFFSET && yDiff < GUIDELINE_OFFSET) {
          candidates.push({
            position: pos,
            xDiff: xDiff,
            yDiff: yDiff,
            offset: offset
          });
        }
      }
    });

    const guides: PieceGuide[] = [];

    // sort guides
    candidates
      .sort((a, b) => {
        if (a.xDiff === b.xDiff) {
          return a.yDiff - b.yDiff;
        }
        return a.xDiff - b.xDiff;
      })
      .forEach((pos) => {
        guides.push({
          position: pos.position,
          offset: Math.max(pos.xDiff, pos.yDiff)
        });
      });

    return guides;
  };
  handleChange = (): void => {
    this.props.onDragMove(this);
  };
  handleMove = (): void => {
    this.props.onDragEnd(this);

    if (this.imageRef.current) {
      this.imageRef.current.setPosition(this.props.position);
    }
  };

  render(): React.ReactNode {
    const img = document.createElement('img');
    img.src = lily;

    return (
      <Image
        ref={this.imageRef}
        offset={{ x: 15, y: 15 }}
        onDragMove={this.handleChange}
        onDragEnd={this.handleMove}
        x={this.state.position.x}
        y={this.state.position.y}
        image={img}
        width={30}
        height={30}
        draggable={true}
      />
    );
  }
}
