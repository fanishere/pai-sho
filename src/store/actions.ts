import {
  MOVE_PIECE,
  PieceGuide,
  Position,
  REMOVE_GUIDES,
  SET_GUIDES
} from './types';
import { createAction } from '@reduxjs/toolkit';

export const movePiece = createAction(
  MOVE_PIECE,
  (pieceName: string, position: Position) => {
    return {
      payload: {
        type: MOVE_PIECE,
        pieceName: pieceName,
        position: position
      }
    };
  }
);

export const setGuides = createAction(SET_GUIDES, (guides: PieceGuide[]) => {
  return {
    payload: {
      type: SET_GUIDES,
      guides
    }
  };
});

export const removeGuides = createAction(REMOVE_GUIDES);
