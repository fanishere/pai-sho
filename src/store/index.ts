import { piecesReducer } from './reducer';
import { configureStore } from '@reduxjs/toolkit';

export const store = configureStore({
  reducer: piecesReducer
});
