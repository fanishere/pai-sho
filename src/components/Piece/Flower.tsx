import { Position } from '../types';
import lily from '../../assets/lily.svg';
import rose from '../../assets/rose.svg';
import jasmine from '../../assets/jasmine.svg';
import chrysan from '../../assets/chrysanthemum.svg';
import rhododendron from '../../assets/rhododendron.svg';
import jade from '../../assets/jade.svg';
import { Piece, PieceProps } from './Piece';

export class Flower extends Piece {
  spaces: number;

  constructor(props: PieceProps) {
    super(props);

    this.spaces = 3;
  }
  getSnapLines = (): Position[] => {
    const lines: Position[] = [];
    if (this.props.position === undefined) {
      return lines;
    }
    for (let i = 0; i <= this.spaces; i++) {
      for (let j = 0; j <= this.spaces; j++) {
        if (i === 0 && j === 0) {
          continue;
        }

        lines.push({
          x: this.props.position.x - this.props.gridWidth * i,
          y: this.props.position.y - this.props.gridWidth * j
        });
        lines.push({
          x: this.props.position.x + this.props.gridWidth * i,
          y: this.props.position.y + this.props.gridWidth * j
        });

        if (i !== 0 && j !== 0) {
          lines.push({
            x: this.props.position.x + this.props.gridWidth * i,
            y: this.props.position.y - this.props.gridWidth * j
          });

          lines.push({
            x: this.props.position.x - this.props.gridWidth * i,
            y: this.props.position.y + this.props.gridWidth * j
          });
        }
      }
    }

    return lines;
  };
}

export class Rose extends Flower {
  constructor(props: PieceProps) {
    super(props);

    const img = document.createElement('img');
    img.src = rose;
    this.image = img;
    this.spaces = 3;
  }
}

export class Jasmine extends Flower {
  constructor(props: PieceProps) {
    super(props);

    const img = document.createElement('img');
    img.src = jasmine;
    this.image = img;
    this.spaces = 3;
  }
}

export class Chrysanthemum extends Flower {
  constructor(props: PieceProps) {
    super(props);

    const img = document.createElement('img');
    img.src = chrysan;
    this.image = img;
    this.spaces = 4;
  }
}

export class Lily extends Flower {
  constructor(props: PieceProps) {
    super(props);

    const img = document.createElement('img');
    img.src = lily;
    this.image = img;
    this.spaces = 4;
  }
}

export class Rhododendron extends Flower {
  constructor(props: PieceProps) {
    super(props);

    const img = document.createElement('img');
    img.src = rhododendron;
    this.image = img;
    this.spaces = 5;
  }
}

export class Jade extends Flower {
  constructor(props: PieceProps) {
    super(props);

    const img = document.createElement('img');
    img.src = jade;
    this.image = img;
    this.spaces = 5;
  }
}
