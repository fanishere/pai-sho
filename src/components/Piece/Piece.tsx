import React, { FunctionComponent, RefObject } from 'react';

import { Image } from 'react-konva';
import { debounce } from 'underscore';
import { PieceGuide, Position, Side } from '../types';
import { useDispatch } from 'react-redux';
import { movePiece, removeGuides, setGuides } from '../../store/actions';
import useImage from 'use-image';
import Konva from 'konva';

export interface PieceProps {
  position: Position | null;
  name: string;
  gridWidth: number;
  side: Side;
  imgSrc: string;
  spaces: number;
}

export const Piece: FunctionComponent<PieceProps> = (props: PieceProps) => {
  const dispatch = useDispatch();

  const imageRef: RefObject<Konva.Image> | null | undefined = React.createRef();
  const getSnapLines = (): Position[] => {
    const lines: Position[] = [];
    if (imageRef.current === null) {
      return lines;
    }
    const currentPosition = imageRef.current.getPosition();
    console.log(currentPosition);
    if (props.position === null) {
      const starterLines = [
        {
          x: 0,
          y: -300
        },
        {
          x: 300,
          y: 0
        },
        {
          x: 0,
          y: 300
        },
        {
          x: -300,
          y: 0
        }
      ];
      return starterLines;
    }
    for (let i = 0; i <= props.spaces; i++) {
      for (let j = 0; j <= props.spaces; j++) {
        if (i === 0 && j === 0) {
          continue;
        }

        lines.push({
          x: props.position.x - props.gridWidth * i,
          y: props.position.y - props.gridWidth * j
        });
        lines.push({
          x: props.position.x + props.gridWidth * i,
          y: props.position.y + props.gridWidth * j
        });

        if (i !== 0 && j !== 0) {
          lines.push({
            x: props.position.x + props.gridWidth * i,
            y: props.position.y - props.gridWidth * j
          });

          lines.push({
            x: props.position.x - props.gridWidth * i,
            y: props.position.y + props.gridWidth * j
          });
        }
      }
    }

    return lines;
  };
  const getGuides = (): PieceGuide[] => {
    const candidates: PieceGuide[] = [];

    if (props.position === undefined) {
      console.log('sdfsfs');
    }

    const snapPositions = getSnapLines();
    snapPositions.forEach((pos) => {
      if (imageRef.current) {
        // TODO check current position is dynamic
        const currentPosition = imageRef.current.getPosition();
        const xDiff = Math.abs(pos.x - currentPosition.x);
        const yDiff = Math.abs(pos.y - currentPosition.y);

        candidates.push({
          position: pos,
          offset: Math.max(xDiff, yDiff)
        });
      }
    });

    // return sorted candidate positions
    return candidates.sort((a, b) => {
      return a.offset - b.offset;
    });
  };

  const handleChange = (): void => {
    const guides = getGuides();
    dispatch(setGuides(guides));
  };
  const handleMove = (): void => {
    const guides = getGuides();
    if (guides.length < 1) {
      return;
    }

    const guide = guides[0];
    let snapPosition = guide.position;
    // offset is too large so snap to original position
    if (guide.offset > 15 && props.position) {
      snapPosition = props.position;
    }

    dispatch(movePiece(props.name, snapPosition));

    if (imageRef.current && props.position) {
      imageRef.current.setPosition(props.position);
    }
    dispatch(removeGuides);
  };

  const image = document.createElement('img');
  image.src = props.imgSrc;
  const [newImg] = useImage(props.imgSrc);

  let x = 300;
  let y = 0;
  if (props.position != undefined) {
    x = props.position.x;
    y = props.position.y;
  }

  return (
    <Image
      image={newImg}
      ref={imageRef}
      offset={{ x: 15, y: 15 }}
      onDragMove={debounce(handleChange, 25, true)}
      onDragEnd={handleMove}
      x={x}
      y={y}
      width={30}
      height={30}
      draggable={true}
      stroke={props.side}
    ></Image>
  );
};
