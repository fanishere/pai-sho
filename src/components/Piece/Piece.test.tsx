/* eslint-disable @typescript-eslint/no-non-null-assertion */
import * as React from 'react';
import { Piece, PieceProps } from './Piece';
import { mount, ReactWrapper, shallow } from 'enzyme';
import { GameState } from '../../store/types';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import rose from '../../assets/rose.svg';

describe('Mounted tests', () => {
  let wrapper: ReactWrapper;
  let piece: ReactWrapper<PieceProps>;
  const initialGameState: GameState = {
    piecePositions: [],
    guidesVisible: []
  };

  beforeEach(() => {
    const middlewares: any = [];
    const mockStore = configureStore(middlewares);
    const store = mockStore(initialGameState);

    wrapper = mount(
      <Provider store={store}>
        <Piece
          name={'rose'}
          gridWidth={20}
          side={'red'}
          imgSrc={rose}
          spaces={3}
          position={{ x: 1, y: 0 }}
        />
      </Provider>
    );
    piece = wrapper.find(Piece);
  });

  describe('Test render', () => {
    it('renders piece with props', () => {
      expect(piece).toMatchSnapshot();
    });

    it('renders with the props set', () => {
      expect(piece.props().gridWidth).toEqual(20);
      expect(piece.props().position).toEqual({ x: 1, y: 0 });
      expect(piece.props().name).toEqual('rose');
    });
  });
});
