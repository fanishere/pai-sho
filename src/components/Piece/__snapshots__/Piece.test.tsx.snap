// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mounted tests Test render renders piece with props 1`] = `
<Piece
  gridWidth={20}
  imgSrc="rose.svg"
  name="rose"
  position={
    Object {
      "x": 1,
      "y": 0,
    }
  }
  side="red"
  spaces={3}
>
  <Image
    draggable={true}
    height={30}
    offset={
      Object {
        "x": 15,
        "y": 15,
      }
    }
    onDragEnd={[Function]}
    onDragMove={[Function]}
    stroke="red"
    width={30}
    x={1}
    y={0}
  />
</Piece>
`;
